File: code_snippets/java/src/main/java/develop/RetryRunService.java
package develop;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.common.RetryPolicy;
import dev.restate.sdk.common.TerminalException;
import java.time.Duration;

@Service
public class RetryRunService {

  @Handler
  public String myHandler(Context ctx, String greeting) {

    // <start_here>
    try {
      // <mark_1>
      RetryPolicy myRunRetryPolicy =
          RetryPolicy.defaultPolicy()
              .setInitialDelay(Duration.ofMillis(500))
              .setExponentiationFactor(2)
              .setMaxDelay(Duration.ofSeconds(10))
              .setMaxAttempts(10)
              .setMaxDuration(Duration.ofMinutes(5));
      // </mark_1>
      ctx.run("my-run", myRunRetryPolicy, () -> writeToOtherSystem());
    } catch (TerminalException e) {
      // Handle the terminal error after retries exhausted
      // For example, undo previous actions (see sagas guide) and
      // propagate the error back to the caller
    }
    // <end_here>

    return greeting + "!";
  }

  private void writeToOtherSystem() {}
}

File: code_snippets/kotlin/src/main/kotlin/develop/RetryRunService.kt
package develop

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.kotlin.Context
import dev.restate.sdk.kotlin.RetryPolicy
import dev.restate.sdk.kotlin.runBlock
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

@Service
class RetryRunService {

  @Handler
  suspend fun retryRun(ctx: Context, greeting: String): String {

    // <start_here>
    try {
      // <mark_1>
      val myRunRetryPolicy =
          RetryPolicy(
              initialDelay = 5.seconds,
              exponentiationFactor = 2.0f,
              maxDelay = 60.seconds,
              maxAttempts = 10,
              maxDuration = 5.minutes)
      // </mark_1>
      ctx.runBlock("write", myRunRetryPolicy) { writeToOtherSystem() }
    } catch (e: TerminalException) {
      // Handle the terminal error after retries exhausted
      // For example, undo previous actions (see sagas guide) and
      // propagate the error back to the caller
      throw e
    }
    // <end_here>

    return "$greeting!"
  }

  private fun writeToOtherSystem() {
    TODO("Not yet implemented")
  }
}
